### PR TITLE
Fix efile_allocate for Mac

### PR DESCRIPTION
**Pull request by indocomsoft**
https://bugs.erlang.org/browse/ERL-1042

### CODE REVIEW

**Comment by jhogberg** [comment]
(Continuing the discussion from [ERL-1042](https://bugs.erlang.org/browse/ERL-1042]))

> Should I keep the current behaviour on Linux?
> Otherwise, I can standardise the behaviour to change the file size as well, in which case I will also change the Linux-specific part of the code to have mode be 0 instead of FALLOC_FL_KEEP_SIZE, and update the documentation

Yes, we don't like introducing behavioral changes in patches but it should be changed on `master` later on as we strive to have the same behavior on all platforms. It should've been this way to begin with in my opinion.

Can you squash and rebase the fix to `maint`, and expand the test in `file_SUITE` to verify that the file size is updated? We can settle for only testing this on the platforms that we know update it at the moment.

**Comment by indocomsoft** [comment]
I am still trying to get the test to fail when I revert the fix.

**Comment by indocomsoft** [comment]
@jhogberg I think I'm finally able to get the test to fail when I revert the fix.

Should I squash everything into 1 commit?

File: lib/kernel/test/file_SUITE.erl [hunk 1]
```diff
@@ -2036,6 +2038,24 @@ allocate_and_assert(Fd, Offset, Length) ->
...
+        {unix, linux} ->
+            {skip, "file:allocate/3 on Linux does not change file size"};
+
+        _ ->
+            PrivDir = proplists:get_value(priv_dir, Config),
+            Allocate = filename:join(PrivDir, atom_to_list(?MODULE)++"_allocate_file"),
+
+            {ok, Fd} = ?FILE_MODULE:open(Allocate, [write]),
+            ok = ?FILE_MODULE:allocate(Fd, 0, 1024),
+            {ok, 1024} = ?FILE_MODULE:position(Fd, eof)
```
Comment (jhogberg) [inline lib/kernel/test/file_SUITE.erl]: The file ought to be closed at the end.

**Comment by indocomsoft** [comment]
Fixed! I also followed the other tests in calling `flush()` and then returning `ok`

**Comment by jhogberg** [comment]
Looks good, if all goes well I'll merge it after `22.1` is released. :)

**Comment by indocomsoft** [comment]
Great! Looking forward to the `22.1` release.

Should I also create a PR to change the Linux behaviour based off `master` after this PR is merged?

**Comment by jhogberg** [comment]
> Should I also create a PR to change the Linux behaviour based off `master` after this PR is merged?

Feel free!

**Comment by jhogberg** [comment]
Merged, thanks again for the PR!

**Pull request merged**